#12068: Fix incorrect env handling in spawnProcess on Posix when posix_spawnp is used.

p12tic opened the pull request:
Fixes #12068. This is a regression starting at Twisted 23.8.0.

[Documentation on reactor.spawnProcess](https://docs.twisted.org/en/stable/api/twisted.internet.interfaces.IReactorProcess.html) says the following about env parameter:

```env is None:  On POSIX: pass os.environ```

This was not correctly handled in `os.posix_spawnp()` code path.

## Scope and purpose

This PR is just a simple one line fix to address #12068. Tests for environment passing on Posix have been added.

p12tic commented:
please review

adiroiban commented:
Thanks for the PR!

Not sure what to do here.

I think that we should fix the documentation instead.

That is, the bug lies in the documentation and not in the code :)

This change was introduced 2 years ago and is present in a public release.

By doing this change, we will break backward compatibility.

The initial change from 16 years ago was using `os.execvpe`

https://github.com/twisted/twisted/blame/dee676b040dd38b847ea6fb112a712cb5e119490/src/twisted/internet/process.py#L468

My understanding is that [execvpe](https://linux.die.net/man/3/execvpe) always requires an explicit environment, and it will not copy the environment from the parent process.

Regards

glyph commented:
> Not sure what to do here.
> 
> I think that we should fix the documentation instead.
> 
> That is, the bug lies in the documentation and not in the code :)
> 
> This change was introduced 2 years ago and is present in a public release.
> 
> By doing this change, we will break backward compatibility.
> 
> The initial change from 16 years ago was using `os.execvpe`

Normally I would agree, except the circumstances under which `posix_spawn` is used are not entirely under control of the user, and this introduces an inconsistency *within* twisted's behaviors.

In this specific case, I think the least impactful thing is to fix the code, rather than the docs.

But perhaps we should shoot this over to the mailing list for a tiebreaker?

p12tic commented:
> This change was introduced 2 years ago and is present in a public release.

This change of behavior was actually released in 23.8.0 by enabling use of `posix_spawnp`.

The regression is actually pretty bad. It silently drops environment variables only under certain conditions (e.g. when `path` argument of `spawnProcess` is not specified). Which means that most of the uses of `spawnProcess` in a certain codebase will work, except a few which depend on environment variables and satisfy the conditions for using `posix_spawnp`. Additionally, `spawnProcess` is something that's usually mocked in unit tests. Which means that unit tests wouldn't catch the failure. Only full end to end or manual testing could possibly catch it. This naturally has much worse test coverage, which means that failures are likely to be unnoticed for quite some time.

This is how the bug was discovered in the first place - by investigating a regression in a downstream project caused by updating Twisted.

> In this specific case, I think the least impactful thing is to fix the code, rather than the docs.

The code fix is needed in either case, because if we fix the docs, we need to adjust the behavior of how `spawnProcess` works in the case when regular `fork` is used.

p12tic commented:
> My understanding is that [execvpe](https://linux.die.net/man/3/execvpe) always requires an explicit environment, and it will not copy the environment from the parent process.

execvpe uses os.environ when env argument is None. See here: https://github.com/python/cpython/blob/3.12/Lib/os.py#L594

adiroiban commented on src/twisted/test/test_process.py:
```diff
@@ -1199,6 +1199,86 @@ def checkTerminated(ignored):
 
         return ended
 
+    def _testEnvironment(self, spawnKwargs, expectedEnv):
```
Maybe this is a better name and the docstring exlpain when it's ok to reuse this code in other tests.

This is testing code so I don't think that we need to worry about private methods.
```suggestion
    def checkSpawnProcessEnvironment(self, spawnKwargs, expectedEnv):
        """
        Shared code for testing the environment variables
        present in the spawned process.

        The spawned process serializes its environ to stderr which is checked against
        os.environ of the calling process.
        """
```

p12tic replied:
Makes sense, thanks.

adiroiban commented on src/twisted/test/test_process.py:
```diff
@@ -1199,6 +1199,86 @@ def checkTerminated(ignored):
+
+    def test_environmentPosixSpawnpEnvNotSet(self):
+        """
+        Spawned process serializes its environ to stderr which is checked against
+        os.environ of the calling process.
```
I am not sure if this is the right docstring.
This looks like the docstring for `_testEnvironment`

I was expecting for the docstring to describe the specific usage scenario for the API.

```suggestion
        An empty environment is passed to the spawned process, when the default value of the C{env} is used.
        That is, when the C{env} argument is not explicitly set.
```

adiroiban commented on src/twisted/test/test_process.py:
```diff
@@ -1199,6 +1199,86 @@ def checkTerminated(ignored):
+
+    def test_environmentForkEnvNotSet(self):
+        """
+        Spawned process serializes its environ to stderr which is checked against
+        os.environ of the calling process.
```
This is my undestanding of this test.

It's not clear why `os.fork` and not `os.posix_spawn` is expected to be used in this test. 

```suggestion
        An empty environment is passed to the spawned process, when the default value of the C{env} is used.
        Passing other argument to the spawned process will
        not affect the environment.
```

p12tic replied:
I've now reused reactor._neverUseSpawn which is used in other tests. This should make the code much clearer compared to setting `path` argument to spawnProcess to something that is not supported by `posix_spawnp`.

adiroiban commented on src/twisted/test/test_process.py:
```diff
@@ -1199,6 +1199,86 @@ def checkTerminated(ignored):
+
+    def test_environmentPosixSpawnpEnvNone(self):
+        """
+        Spawned process serializes its environ to stderr which is checked against
+        os.environ of the calling process.
```
```suggestion
        The parent environment is passed to the spawned process, when the C{env} argument is set to C{None}.
```

p12tic replied:
Fixed.

adiroiban commented on src/twisted/test/test_process.py:
```diff
@@ -1199,6 +1199,86 @@ def checkTerminated(ignored):
+
+    def test_environmentPosixSpawnpEnvCustom(self):
+        """
+        Spawned process serializes its environ to stderr which is checked against
+        os.environ of the calling process.
```
```suggestion
        The explicit C{env} environment is passed to the spawned process.
```

p12tic replied:
Fixed.

adiroiban commented on src/twisted/test/test_process.py:
```diff
@@ -1199,6 +1199,86 @@ def checkTerminated(ignored):
+
+    def test_environmentForkEnvNone(self):
+        """
+        Spawned process serializes its environ to stderr which is checked against
```
I am not sure why `fork` is expected to be triggered by this test.

I think that a dedicated docstring could help identity what is tested here and answer the question: "Why we need this test?"

p12tic replied:
Fixed.

adiroiban commented on src/twisted/test/test_process.py:
```diff
@@ -1199,6 +1199,86 @@ def checkTerminated(ignored):
+
+    def test_environmentForkEnvCustom(self):
+        """
+        Spawned process serializes its environ to stderr which is checked against
```
Same. Not sure why `fork` is expected to be used here.

Please consider adding a dedicated docstring to describe this test.

p12tic replied:
Fixed.

adiroiban commented on src/twisted/newsfragments/12068.bugfix:
```diff
@@ -0,0 +1,2 @@
+Fixed invalid handling of ``env=None`` in ``reactor.spawnProcess`` on Posix
+when os.posix_spawnp() is being used.
```
```suggestion
twisted.internet.process.Process, used by ``reactor.spawnProcess``, now copies the parent environment when the `env=None` argument is passed on Posix systems.
```

Is there a case of a supported OS when os.posix_spawnp is not used on Posix (Linux and macOS) ?

I guess that the fork code can be removed / deprecated.
Are there still posix OSes without `posix_spawnp` ?

p12tic replied:
Unfortunately posix_spawnp does not support changing the current directory, uid and gid. Setting current directory has been accepted to POSIX (https://www.austingroupbugs.net/view.php?id=1208), but for uid and gid we'll need to use fork+execvpe for a long time it seems.

adiroiban replied:
Many thanks for the info.
Just a question.

Do you think it is worth updating the current documentation to explain that posix_spawnp is used when possible, and if you use path and uid/gid the good old fork + execvpe is used ?

glyph replied:
> Do you think it is worth updating the current documentation to explain that posix_spawnp is used when possible, and if you use path and uid/gid the good old fork + execvpe is used ?

If so, not in this PR :)

p12tic replied:
> Do you think it is worth updating the current documentation to explain that posix_spawnp is used when possible, and if you use path and uid/gid the good old fork + execvpe is used ?

I think this is useful addition. Usage of os.fork is user visible because new python throws a warning, so end users may want to adjust their code to not use it (e.g. invoke spawnProcess differently).

I've opened https://github.com/twisted/twisted/pull/12072.

p12tic replied:
@adiroiban I think it's important to mention that only posix_spawnp code path is being affected in the release note. I've used your wording as baseline and adjusted it to include posix_spawnp note.

adiroiban reviewed (changes requested):
Many thanks, Povilas for the PR and the explanations and investigation.

With the latest info, I also think that is best to fix the code.

I left a few minor comments regarding the scope/purpose of each test.

I tried to read the code, but I can't find why fork is expected to be used in those tests.

Please check my commend and add your feedback.

Other than that, this can be merged.

Once merged, I will try to trigger a new release.

Thanks again.

p12tic commented:
@adiroiban I addressed your comments.

p12tic commented:
please review

p12tic committed:
Fix incorrect env handling in spawnProcess on Posix

Documentation on reactor.spawnProcess says the following about env
parameter:

env is None:  On POSIX: pass os.environ

This was not correctly handled in os.posix_spawnp() code path.

adiroiban commented on src/twisted/test/test_process.py:
```diff
@@ -1199,6 +1199,115 @@ def checkTerminated(ignored):
+        p = Accumulator()
+        d = p.endedDeferred = defer.Deferred()
+
+        reactor._neverUseSpawn = not usePosixSpawnp
```
Instead of using the global reactor,
maybe we can instantiate a separate reactor and then we will not have to care about the cleanup.

The suggestion is just an idea... the import should be at the start of the file.

```suggestion
        from twisted.internet.selectreactor import SelectReactor()
        supportingReactor = SelectReactor()
        supportingReactor._neverUseSpawn = not usePosixSpawnp
```

I think that is important not to modify the global / singleton instanced during a test.

exarkun replied:
This isn't a safe way to do this.  Instead, there's a whole framework in twisted/internet/test/ for using per-test reactors (including many tests for `IReactorProcess`).

adiroiban replied:
Thanks for the info.

My suggestion is not ok as it only guarantees that the `spawnProcess` will work for the `SelectReactor` 

The test is not executed for all the other reactors.

I guess that these tests needs to be moved to https://github.com/twisted/twisted/blob/trunk/src/twisted/internet/test/test_process.py#L106-L113

and use `reactor = self.buildReactor()`

In this way, the tests are wrote once, but executed for any supported reactor.

glyph replied:
> I guess that these tests needs to be moved to https://github.com/twisted/twisted/blob/trunk/src/twisted/internet/test/test_process.py#L106-L113

Yes, exactly.

`twisted.test` is largely a trap for contributors at this point. We need to more clearly signpost the good vs. the bad ways to write tests and try to direct people to copy from the good examples (or at least, the better ones) and avoid these antipatterns.

p12tic replied:
Thanks for suggestions. I've applied them all.

adiroiban commented on src/twisted/test/test_process.py:
```diff
@@ -1199,6 +1199,115 @@ def checkTerminated(ignored):
 
         return ended
 
+    def checkSpawnProcessEnvironment(self, usePosixSpawnp, spawnKwargs, expectedEnv):
```
I find it hard to read this code and know that the `False` means "use fork"

```python
return self.checkSpawnProcessEnvironment(
            False,
            {"env": {"MYENV1": "myvalue1"}},
            {"MYENV1": "myvalue1"},
        )
```

Just a minor suggestion.

We can also merge the current test code as it is 

The suggestion is just an idea, not the actual suggestion to commit.

```suggestion
    def checkSpawnProcessEnvironmentWithFork():
        return self._checkSpawnProcessEnvironment(self, spawnKwargs, expectedEnv, usePosixSpawnp=False)
    def checkSpawnProcessEnvironmentWithPosixSpawnp():
        return self._checkSpawnProcessEnvironment(self, spawnKwargs, expectedEnv, usePosixSpawnp=True)
    def _checkSpawnProcessEnvironment(self, spawnKwargs, expectedEnv, usePosixSpawnp):
```

p12tic replied:
Thanks, applied.

adiroiban reviewed (dismissed):
Many thanks for the PR.

It looks good.

My only blocker for this PR is trying to use a dedicated reactor instance, rather than using the global reactor.

Maybe it's a bad suggestion :)

See my inline comments. Thanks!

p12tic commented:
please review

p12tic commented:
@adiroiban 

Thanks a lot for your comments. I think I addressed all comments.

glyph reviewed (approved):
Looks great. Thanks for integrating all that feedback.
